File: HerdMe/Services/AppUpdateManager.swift
import CryptoKit
import Foundation

struct AppUpdateDownloadURLs: Codable, Equatable, Sendable {
    let macOS: URL?
    let windowsX64: URL?
}

struct AppUpdateRelease: Codable, Equatable, Sendable {
    let version: String
    let build: Int
    let channel: String
    let notes: String
    let downloadURL: URL?
    let downloadURLs: AppUpdateDownloadURLs?

    init(
        version: String,
        build: Int,
        channel: String,
        notes: String,
        downloadURL: URL?,
        downloadURLs: AppUpdateDownloadURLs? = nil
    ) {
        self.version = version
        self.build = build
        self.channel = channel
        self.notes = notes
        self.downloadURL = downloadURL
        self.downloadURLs = downloadURLs
    }

    var platformDownloadURL: URL? {
        downloadURLs?.macOS ?? downloadURL
    }
}

struct AppUpdateManifest: Codable, Equatable, Sendable {
    let releases: [AppUpdateRelease]
}

struct AppUpdateSignedEnvelope: Codable, Equatable, Sendable {
    static let algorithm = "ECDSA_P256_SHA256"

    let algorithm: String
    let payload: String
    let signature: String
}

enum AppUpdateResult: Equatable, Sendable {
    case upToDate(version: String)
    case available(AppUpdateRelease)
}

enum AppUpdateError: LocalizedError, Equatable, Sendable {
    case invalidResponse
    case unsignedRemoteManifest
    case verificationKeyMissing
    case invalidSignature
    case incompletePlatformDownloads
    case noRelease(channel: String)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            "The update service returned an invalid response."
        case .unsignedRemoteManifest:
            "The remote update feed is not signed and was rejected."
        case .verificationKeyMissing:
            "This build does not contain the public key required to verify updates."
        case .invalidSignature:
            "The update feed signature is invalid. The update was rejected."
        case .incompletePlatformDownloads:
            "The signed update feed must contain HTTPS downloads for macOS and Windows x64."
        case let .noRelease(channel):
            "No \(channel.lowercased()) release is available in the update feed."
        }
    }
}

struct AppUpdateManager: Sendable {
    private static let maximumFeedSize = 4 * 1_024 * 1_024

    let feedURL: URL
    let currentVersion: String
    let currentBuild: Int
    let verificationKey: Data?

    init(
        feedURL: URL,
        currentVersion: String,
        currentBuild: Int,
        verificationKey: Data? = nil
    ) {
        self.feedURL = feedURL
        self.currentVersion = currentVersion
        self.currentBuild = currentBuild
        self.verificationKey = verificationKey
    }

    static func configured(bundle: Bundle = .main) -> AppUpdateManager? {
        #if DEBUG
        let environmentURL = ProcessInfo.processInfo.environment["HERDME_UPDATE_FEED_URL"]
            .flatMap(URL.init(string:))
        let environmentKey = ProcessInfo.processInfo.environment["HERDME_UPDATE_PUBLIC_KEY"]
            .flatMap(Self.decodePublicKey)
        #else
        let environmentURL: URL? = nil
        let environmentKey: Data? = nil
        #endif
        let configuredURL = (bundle.object(forInfoDictionaryKey: "HerdMeUpdateFeedURL") as? String)
            .flatMap(Self.httpsURL)
        let bundledFeedURL = bundle.url(forResource: "release-feed-url", withExtension: "txt")
            .flatMap { try? String(contentsOf: $0, encoding: .utf8) }
            .flatMap(Self.httpsURL)
        guard let feedURL = environmentURL
                ?? configuredURL
                ?? bundledFeedURL
                ?? bundle.url(forResource: "release-manifest", withExtension: "json") else {
            return nil
        }
        let version = bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
            ?? "0.0.0"
        let build = Int(bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "0") ?? 0
        let configuredKey = (bundle.object(forInfoDictionaryKey: "HerdMeUpdatePublicKey") as? String)
            .flatMap(Self.decodePublicKey)
        let bundledKey = bundle.url(forResource: "release-public-key", withExtension: "txt")
            .flatMap { try? String(contentsOf: $0, encoding: .utf8) }
            .flatMap(Self.decodePublicKey)
        return AppUpdateManager(
            feedURL: feedURL,
            currentVersion: version,
            currentBuild: build,
            verificationKey: environmentKey ?? configuredKey ?? bundledKey
        )
    }

    func check(channel: String) async throws -> AppUpdateResult {
        let data: Data
        if feedURL.isFileURL {
            data = try Data(contentsOf: feedURL)
        } else {
            let (remoteData, response) = try await ManagedDownloadClient.data(
                from: feedURL,
                session: Self.remoteSession
            )
            guard let httpResponse = response as? HTTPURLResponse,
                  200..<300 ~= httpResponse.statusCode else {
                throw AppUpdateError.invalidResponse
            }
            data = remoteData
        }
        guard data.count <= Self.maximumFeedSize else { throw AppUpdateError.invalidResponse }
        let manifest = try Self.decodeManifest(
            data,
            requiresSignature: !feedURL.isFileURL,
            publicKey: verificationKey
        )
        let selectedChannel = channel.lowercased()
        let acceptedChannels = selectedChannel == "beta"
            ? Set(["stable", "beta"])
            : Set(["stable"])
        let eligibleReleases = manifest.releases
            .filter { acceptedChannels.contains($0.channel.lowercased()) }
            .sorted(by: Self.isOrderedAfter)
        guard let latest = eligibleReleases.first else {
            throw AppUpdateError.noRelease(channel: channel)
        }
        return Self.isNewer(latest, than: currentVersion, build: currentBuild)
            ? .available(latest)
            : .upToDate(version: currentVersion)
    }

    private static func isOrderedAfter(_ lhs: AppUpdateRelease, _ rhs: AppUpdateRelease) -> Bool {
        let order = VersionComparison.compare(lhs.version, rhs.version)
        return order == .orderedSame ? lhs.build > rhs.build : order == .orderedDescending
    }

    private static func isNewer(
        _ release: AppUpdateRelease,
        than version: String,
        build: Int
    ) -> Bool {
        let order = VersionComparison.compare(release.version, version)
        return order == .orderedDescending || order == .orderedSame && release.build > build
    }

    static func decodeManifest(
        _ data: Data,
        requiresSignature: Bool,
        publicKey: Data?
    ) throws -> AppUpdateManifest {
        if let envelope = try? JSONDecoder().decode(AppUpdateSignedEnvelope.self, from: data) {
            return try verify(envelope: envelope, publicKey: publicKey)
        }
        guard !requiresSignature else {
            throw AppUpdateError.unsignedRemoteManifest
        }
        return try validate(
            JSONDecoder().decode(AppUpdateManifest.self, from: data),
            requiresPlatformDownloads: false
        )
    }

    static func verify(
        envelope: AppUpdateSignedEnvelope,
        publicKey: Data?
    ) throws -> AppUpdateManifest {
        guard envelope.algorithm == AppUpdateSignedEnvelope.algorithm,
              let payload = Data(base64Encoded: envelope.payload),
              let signatureData = Data(base64Encoded: envelope.signature) else {
            throw AppUpdateError.invalidSignature
        }
        guard let publicKey else { throw AppUpdateError.verificationKeyMissing }
        do {
            let key = try P256.Signing.PublicKey(x963Representation: publicKey)
            let signature = try P256.Signing.ECDSASignature(derRepresentation: signatureData)
            guard key.isValidSignature(signature, for: payload) else {
                throw AppUpdateError.invalidSignature
            }
        } catch let error as AppUpdateError {
            throw error
        } catch {
            throw AppUpdateError.invalidSignature
        }
        guard payload.count <= maximumFeedSize else { throw AppUpdateError.invalidResponse }
        return try validate(
            JSONDecoder().decode(AppUpdateManifest.self, from: payload),
            requiresPlatformDownloads: true
        )
    }

    private static func validate(
        _ manifest: AppUpdateManifest,
        requiresPlatformDownloads: Bool
    ) throws -> AppUpdateManifest {
        let baseIsValid = manifest.releases.allSatisfy { release in
            release.build >= 0
                && ["stable", "beta"].contains(release.channel.lowercased())
                && isValidVersion(release.version)
                && isValidHTTPSURL(release.downloadURL, required: false)
                && (release.downloadURLs == nil || hasCompletePlatformDownloads(release))
        }
        guard baseIsValid else { throw AppUpdateError.invalidResponse }
        if requiresPlatformDownloads,
           !manifest.releases.allSatisfy(hasCompletePlatformDownloads) {
            throw AppUpdateError.incompletePlatformDownloads
        }
        return manifest
    }

    private static func hasCompletePlatformDownloads(_ release: AppUpdateRelease) -> Bool {
        guard let downloads = release.downloadURLs,
              downloads.macOS != downloads.windowsX64 else {
            return false
        }
        return isValidHTTPSURL(downloads.macOS, required: true)
            && isValidHTTPSURL(downloads.windowsX64, required: true)
    }

    private static func isValidHTTPSURL(_ url: URL?, required: Bool) -> Bool {
        guard let url else { return !required }
        return url.scheme == "https" && url.host != nil
    }

    private static func isValidVersion(_ value: String) -> Bool {
        VersionComparison.isSemanticVersion(value)
    }

    private static func decodePublicKey(_ value: String) -> Data? {
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let key = Data(base64Encoded: normalized), key.count == 65, key.first == 4 else {
            return nil
        }
        return key
    }

    private static func httpsURL(_ value: String) -> URL? {
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = URL(string: normalized), url.scheme == "https", url.host != nil else {
            return nil
        }
        return url
    }

    private static let remoteSession: URLSession = {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 20
        configuration.timeoutIntervalForResource = 60
        configuration.waitsForConnectivity = false
        return URLSession(configuration: configuration)
    }()
}

File: Windows/HerdMe.Windows/Services/AppUpdateManager.cs
using System.Reflection;
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HerdMe.Windows.Services;

public sealed record AppUpdateDownloadUrls(
    [property: JsonPropertyName("macOS")] string? MacOS,
    [property: JsonPropertyName("windowsX64")] string? WindowsX64
);

public sealed record AppUpdateRelease(
    [property: JsonPropertyName("version")] string Version,
    [property: JsonPropertyName("build")] int Build,
    [property: JsonPropertyName("channel")] string Channel,
    [property: JsonPropertyName("notes")] string Notes,
    [property: JsonPropertyName("downloadURL")] string? DownloadUrl,
    [property: JsonPropertyName("downloadURLs")] AppUpdateDownloadUrls? DownloadUrls = null
)
{
    [JsonIgnore]
    public string? PlatformDownloadUrl => DownloadUrls?.WindowsX64 ?? DownloadUrl;
}

public sealed record AppUpdateCheck(
    string CurrentVersion,
    int CurrentBuild,
    AppUpdateRelease? AvailableRelease
)
{
    public bool IsAvailable => AvailableRelease is not null;
}

public sealed class AppUpdateManager
{
    private const string SignatureAlgorithm = "ECDSA_P256_SHA256";
    private const int MaximumFeedSize = 4 * 1024 * 1024;
    private static readonly HttpClient HttpClient = new()
    {
        Timeout = TimeSpan.FromSeconds(60)
    };
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    private readonly string feedLocation;
    private readonly string currentVersion;
    private readonly int currentBuild;
    private readonly byte[]? verificationKey;

    public AppUpdateManager(
        string feedLocation,
        string currentVersion,
        int currentBuild,
        string? publicKey = null
    )
    {
        this.feedLocation = feedLocation;
        this.currentVersion = currentVersion;
        this.currentBuild = currentBuild;
        verificationKey = DecodePublicKey(publicKey);
    }

    public static AppUpdateManager Configured(Assembly? assembly = null)
    {
        assembly ??= Assembly.GetEntryAssembly() ?? typeof(AppUpdateManager).Assembly;
        var assemblyVersion = assembly.GetName().Version ?? new Version(0, 0, 0, 0);
        var informationalVersion = assembly
            .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?
            .InformationalVersion
            .Split('+', 2)[0];
        var version = string.IsNullOrWhiteSpace(informationalVersion)
            ? $"{assemblyVersion.Major}.{assemblyVersion.Minor}.{Math.Max(assemblyVersion.Build, 0)}"
            : informationalVersion;
        var build = Math.Max(assemblyVersion.Revision, 0);
        string? feed = null;
        string? publicKey = null;
#if DEBUG
        feed = Environment.GetEnvironmentVariable("HERDME_UPDATE_FEED_URL");
        publicKey = Environment.GetEnvironmentVariable("HERDME_UPDATE_PUBLIC_KEY");
#endif
        if (string.IsNullOrWhiteSpace(feed))
        {
            var feedConfigurationPath = Path.Combine(AppContext.BaseDirectory, "release-feed-url.txt");
            var configuredFeed = File.Exists(feedConfigurationPath)
                ? File.ReadAllText(feedConfigurationPath).Trim()
                : null;
            feed = Uri.TryCreate(configuredFeed, UriKind.Absolute, out var configuredUri)
                && configuredUri.Scheme == Uri.UriSchemeHttps
                    ? configuredFeed
                    : Path.Combine(AppContext.BaseDirectory, "release-manifest.json");
        }
        if (string.IsNullOrWhiteSpace(publicKey))
        {
            var publicKeyPath = Path.Combine(AppContext.BaseDirectory, "release-public-key.txt");
            if (File.Exists(publicKeyPath))
            {
                publicKey = File.ReadAllText(publicKeyPath);
            }
        }
        return new AppUpdateManager(feed, version, build, publicKey);
    }

    public async Task<AppUpdateCheck> CheckAsync(
        string channel,
        CancellationToken cancellationToken = default
    )
    {
        var data = await ReadFeedAsync(cancellationToken);
        var manifest = DecodeManifest(data, IsRemoteFeed(feedLocation));
        var normalizedChannel = channel.Equals("Beta", StringComparison.OrdinalIgnoreCase)
            ? "beta"
            : "stable";
        var acceptedChannels = normalizedChannel == "beta"
            ? new HashSet<string>(["stable", "beta"], StringComparer.OrdinalIgnoreCase)
            : new HashSet<string>(["stable"], StringComparer.OrdinalIgnoreCase);
        var latest = manifest.Releases
            .Where(release => acceptedChannels.Contains(release.Channel))
            .OrderByDescending(release => release, AppUpdateReleaseComparer.Instance)
            .FirstOrDefault()
            ?? throw new InvalidDataException(
                $"No {normalizedChannel} release is available in the update feed."
            );
        var available = IsNewer(latest) ? latest : null;
        return new AppUpdateCheck(currentVersion, currentBuild, available);
    }

    private async Task<byte[]> ReadFeedAsync(CancellationToken cancellationToken)
    {
        if (Uri.TryCreate(feedLocation, UriKind.Absolute, out var uri))
        {
            if (uri.Scheme is "http" or "https")
            {
                using var response = await HttpClient.GetAsync(
                    uri,
                    HttpCompletionOption.ResponseHeadersRead,
                    cancellationToken
                );
                response.EnsureSuccessStatusCode();
                if (response.Content.Headers.ContentLength is > MaximumFeedSize)
                {
                    throw new InvalidDataException("The update feed is too large.");
                }
                await using var input = await response.Content.ReadAsStreamAsync(cancellationToken);
                return await ReadBoundedAsync(input, cancellationToken);
            }
            if (uri.IsFile)
            {
                return await ReadFileAsync(uri.LocalPath, cancellationToken);
            }
        }
        return await ReadFileAsync(feedLocation, cancellationToken);
    }

    private AppUpdateManifest DecodeManifest(byte[] data, bool requireSignature)
    {
        AppUpdateSignedEnvelope? envelope;
        try
        {
            envelope = JsonSerializer.Deserialize<AppUpdateSignedEnvelope>(data, JsonOptions);
        }
        catch (JsonException)
        {
            envelope = null;
        }
        if (envelope is not null
            && envelope.Algorithm is not null
            && envelope.Payload is not null
            && envelope.Signature is not null)
        {
            return VerifyEnvelope(envelope);
        }
        if (requireSignature)
        {
            throw new InvalidDataException("The remote update feed is not signed and was rejected.");
        }
        return ValidateManifest(
            JsonSerializer.Deserialize<AppUpdateManifest>(data, JsonOptions)
                ?? throw new InvalidDataException("The update feed is invalid."),
            requirePlatformDownloads: false
        );
    }

    private AppUpdateManifest VerifyEnvelope(AppUpdateSignedEnvelope envelope)
    {
        if (envelope.Payload is not { } encodedPayload
            || envelope.Signature is not { } encodedSignature)
        {
            throw new InvalidDataException("The update feed signature is invalid.");
        }
        if (!string.Equals(envelope.Algorithm, SignatureAlgorithm, StringComparison.Ordinal)
            || verificationKey is null)
        {
            throw new InvalidDataException(
                verificationKey is null
                    ? "This build does not contain the public key required to verify updates."
                    : "The update feed signature is invalid."
            );
        }
        byte[] payload;
        byte[] signature;
        try
        {
            payload = Convert.FromBase64String(encodedPayload);
            signature = Convert.FromBase64String(encodedSignature);
        }
        catch (FormatException error)
        {
            throw new InvalidDataException("The update feed signature is invalid.", error);
        }

        try
        {
            var parameters = new ECParameters
            {
                Curve = ECCurve.NamedCurves.nistP256,
                Q = new ECPoint
                {
                    X = verificationKey[1..33],
                    Y = verificationKey[33..65]
                }
            };
            using var verifier = ECDsa.Create(parameters);
            if (!verifier.VerifyData(
                    payload,
                    signature,
                    HashAlgorithmName.SHA256,
                    DSASignatureFormat.Rfc3279DerSequence
                ))
            {
                throw new InvalidDataException("The update feed signature is invalid.");
            }
        }
        catch (CryptographicException error)
        {
            throw new InvalidDataException("The update feed signature is invalid.", error);
        }
        if (payload.Length > MaximumFeedSize)
        {
            throw new InvalidDataException("The signed update feed is too large.");
        }
        return ValidateManifest(
            JsonSerializer.Deserialize<AppUpdateManifest>(payload, JsonOptions)
                ?? throw new InvalidDataException("The signed update feed is invalid."),
            requirePlatformDownloads: true
        );
    }

    private static AppUpdateManifest ValidateManifest(
        AppUpdateManifest manifest,
        bool requirePlatformDownloads
    )
    {
        if (manifest.Releases is null)
        {
            throw new InvalidDataException("The update feed has no release collection.");
        }
        foreach (var release in manifest.Releases)
        {
            var channelIsValid = string.Equals(
                    release.Channel,
                    "stable",
                    StringComparison.OrdinalIgnoreCase
                ) || string.Equals(release.Channel, "beta", StringComparison.OrdinalIgnoreCase);
            var legacyDownloadIsValid = IsValidHttpsUrl(release.DownloadUrl, required: false);
            var platformDownloadsAreComplete = HasCompletePlatformDownloads(release);
            if (release.Build < 0
                || string.IsNullOrWhiteSpace(release.Version)
                || !RuntimeVersionComparison.IsSemanticVersion(release.Version)
                || release.Notes is null
                || !channelIsValid
                || !legacyDownloadIsValid
                || release.DownloadUrls is not null && !platformDownloadsAreComplete)
            {
                throw new InvalidDataException("The update feed contains an invalid release.");
            }
            if (requirePlatformDownloads && !platformDownloadsAreComplete)
            {
                throw new InvalidDataException(
                    "The signed update feed must contain HTTPS downloads for macOS and Windows x64."
                );
            }
        }
        return manifest;
    }

    private static bool HasCompletePlatformDownloads(AppUpdateRelease release)
    {
        return release.DownloadUrls is { } downloads
            && TryGetHttpsUri(downloads.MacOS, out var macOSUri)
            && TryGetHttpsUri(downloads.WindowsX64, out var windowsUri)
            && !macOSUri.Equals(windowsUri);
    }

    private static bool IsValidHttpsUrl(string? value, bool required)
    {
        if (string.IsNullOrWhiteSpace(value)) return !required;
        return TryGetHttpsUri(value, out _);
    }

    private static bool TryGetHttpsUri(string? value, out Uri uri)
    {
        return Uri.TryCreate(value, UriKind.Absolute, out uri!)
            && uri.Scheme == Uri.UriSchemeHttps
            && !string.IsNullOrWhiteSpace(uri.Host);
    }

    private static async Task<byte[]> ReadFileAsync(
        string path,
        CancellationToken cancellationToken
    )
    {
        var file = new FileInfo(path);
        if (file.Length > MaximumFeedSize)
        {
            throw new InvalidDataException("The update feed is too large.");
        }
        return await File.ReadAllBytesAsync(path, cancellationToken);
    }

    private static async Task<byte[]> ReadBoundedAsync(
        Stream input,
        CancellationToken cancellationToken
    )
    {
        using var output = new MemoryStream();
        var buffer = new byte[64 * 1024];
        while (true)
        {
            var count = await input.ReadAsync(buffer, cancellationToken);
            if (count == 0) return output.ToArray();
            if (output.Length + count > MaximumFeedSize)
            {
                throw new InvalidDataException("The update feed is too large.");
            }
            output.Write(buffer, 0, count);
        }
    }

    private static byte[]? DecodePublicKey(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        try
        {
            var key = Convert.FromBase64String(value.Trim());
            return key.Length == 65 && key[0] == 4 ? key : null;
        }
        catch (FormatException)
        {
            return null;
        }
    }

    private static bool IsRemoteFeed(string location)
    {
        return Uri.TryCreate(location, UriKind.Absolute, out var uri)
            && uri.Scheme is "http" or "https";
    }

    private bool IsNewer(AppUpdateRelease release)
    {
        var versionOrder = RuntimeVersionComparison.Compare(release.Version, currentVersion);
        return versionOrder > 0 || versionOrder == 0 && release.Build > currentBuild;
    }

    private sealed class AppUpdateReleaseComparer : IComparer<AppUpdateRelease>
    {
        public static AppUpdateReleaseComparer Instance { get; } = new();

        public int Compare(AppUpdateRelease? left, AppUpdateRelease? right)
        {
            if (ReferenceEquals(left, right)) return 0;
            if (left is null) return -1;
            if (right is null) return 1;
            var versionOrder = RuntimeVersionComparison.Compare(left.Version, right.Version);
            return versionOrder != 0 ? versionOrder : left.Build.CompareTo(right.Build);
        }
    }

    private sealed class AppUpdateManifest
    {
        [JsonPropertyName("releases")]
        public List<AppUpdateRelease> Releases { get; init; } = [];
    }

    private sealed class AppUpdateSignedEnvelope
    {
        [JsonPropertyName("algorithm")]
        public string? Algorithm { get; init; }

        [JsonPropertyName("payload")]
        public string? Payload { get; init; }

        [JsonPropertyName("signature")]
        public string? Signature { get; init; }
    }
}
